### PR TITLE
tun: export GSOSplit() for external Device implementers

### PR DESCRIPTION
**Pull request by jwhited**
External implementers of tun.Device may support GSO, and may also be platform-agnostic, e.g. gVisor.

### CODE REVIEW

**Comment by jwhited** [comment]
https://github.com/tailscale/tailscale/pull/12869 is an example usage.

Fuzz tests `Fuzz_handleGRO()` and `Fuzz_GSOSplit()` are running now. I'll add a comment with the number of executions and branches found after they've run for a sufficient amount of time.

**Comment by jwhited** [comment]
`Fuzz_handleGRO()`:

```
fuzz: elapsed: 3h13m5s, execs: 593225797 (48144/sec), new interesting: 22 (total: 194)
PASS
ok      github.com/tailscale/wireguard-go/tun   11585.564s
```

`Fuzz_GSOSplit()`:

```
fuzz: elapsed: 3h13m1s, execs: 17124403376 (1395524/sec), new interesting: 0 (total: 99)
PASS
ok      github.com/tailscale/wireguard-go/tun   11580.631s
```

File: tun/offload.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	ipProtoUDP = 17
+)
+
+// GSOSplit splits packets from 'in' into outBufs[n][outOffset:], writing the
+// size of each element into sizes. It returns the number of buffers populated,
+// and/or an error. Callers may pass an 'in' slice that overlaps with the first
+// element of outBuffers, i.e. &in[0] may be equal to &outBufs[0][outOffset].
+// GSONone is a valid options.GSOType regardless of the value of
+// options.NeedsCsum. Length of each outBufs element must be greater than or
+// equal to the length of 'in'.
```
Comment (raggi) [inline tun/offload.go]: add a note that if outbufs are not long enough, the output will be silently truncated, as i believe that's the semantic we've got going on

File: tun/offload.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	tcpFlagACK uint8 = 0x10
+)
+
+const (
+	// defined here in order to avoid importation of any platform-specific pkgs
+	ipProtoTCP = 6
+	ipProtoUDP = 17
+)
+
+// GSOSplit splits packets from 'in' into outBufs[n][outOffset:], writing the
```
Comment (raggi) [inline tun/offload.go]: small issue: n isn't actually defined

**Review by raggi — APPROVED** [review]

File: tun/offload.go (hunk 1 above)
Comment (jwhited) [reply tun/offload.go]: updated in https://github.com/tailscale/wireguard-go/pull/29/commits/f6e967f590e2521856356ec61060980984aba26d

File: tun/offload.go (hunk 2 above)
Comment (jwhited) [reply tun/offload.go]: updated in https://github.com/tailscale/wireguard-go/pull/29/commits/209d02a97f4b8be2040052e6581ebc8d6b30ca5a

**Pull request merged**